docs/Settings.md
# Settings

Settings can be customized extensively, and even nested in advanced components. The values for some settings can be customized as well.

## Display Order

You can customize the order of the actions in the `reflowOrder` or `fxlOrder` arrays, and remove them as well if you don’t want to expose some. 

Enum `SettingsKeys` is provided to keep things consistent across the entire codebase.

For instance:

```
settings: {
  ...
  reflowOrder: [
    SettingsKeys.zoom,
    SettingsKeys.fontFamily,
    SettingsKeys.theme,
    SettingsKeys.lineHeight,
    SettingsKeys.layout,
    SettingsKeys.columns
  ],
  fxlOrder: [
    SettingsKeys.zoom,
    SettingsKeys.theme,
    SettingsKeys.columns
  ]
}
```

Note that if you are using a standalone component in an Advanced component (either in `main` or `displayOrder`), it will be filtered so that it is not rendered twice.

### Standalone components

All settings components are standalone by default, which means you can organise them as you see fit to build your own Settings panel.

### Advanced components

`SettingsKey.text` and `SettingsKeys.spacing` are two advanced components in which you can nest other components.

Enums `TextSettingsKeys` and `SpacingSettingsKeys` list which components can be nested in these two.

When used, a button will be added to access the nested components.

## Text (optional)

The text object is responsible for the advanced Text Component, which provides an extra container to display more options.

### Main (optional)

The `main` property accepts an array of `TextSettingsKeys`. These components will be displayed in the SettingsPanel, with a button to access the components in `displayOrder`.

If all nestable components are listed in `main`, then the Text component behaves as if all its nested components are standalone, and will not create a button to access them – as they are already accessible.

### DisplayOrder (optional)

The components to display in the “sub-panel”, and their order. Note components listed in `main` will not automatically be added to this array.

## Spacing (optional)

The spacing object is responsible for the advanced Spacing Component, which provides an extra container to display more options.

### Main (optional)

The `main` property accepts an array of `SpacingSettingsKeys`. These components will be displayed in the SettingsPanel, with a button to access the components in `displayOrder`.

If all nestable components are listed in `main`, then the Spacing component behaves as if all its nested components are standalone, and will not create a button to access them – as they are already accessible.

### DisplayOrder (optional)

The components to display in the “sub-panel”, and their order. Note components listed in `main` will not automatically be added to this array.

### LineHeight (optional)

This allows to customize the value for line-heights. It must be a ratio (`number`).

For instance:

```
spacing:
  ... 
  lineHeight: {
    [ReadingDisplayLineHeightOptions.small]: 1.3,
    [ReadingDisplayLineHeightOptions.medium]: 1.5,
    [ReadingDisplayLineHeightOptions.large]: 1.75
  }
}
```

src/models/settings.ts
import { ComponentType, SVGProps } from "react";
import { ReadingDisplayAlignOptions, ReadingDisplayLineHeightOptions, RSLayoutStrategy } from "./layout";
import { ThemeKeys } from "./theme";
import { PressEvent, TooltipProps } from "react-aria-components";
import { SheetHeaderVariant } from "./sheets";

export enum SettingsContainerKeys {
  initial = "initial",
  text = "text",
  spacing = "spacing"
}

export enum SettingsKeys {
  align = "align",
  columns = "columns",
  fontFamily = "fontFamily",
  fontWeight = "fontWeight",
  hyphens = "hyphens",
  layout = "layout",
  letterSpacing = "letterSpacing",
  lineHeight = "lineHeight",
  normalizeText = "normalizeText",
  paraIndent = "paraIndent",
  paraSpacing = "paraSpacing",
  publisherStyles = "publisherStyles",
  spacing = "spacing",
  text = "text",
  theme = "theme",
  wordSpacing = "wordSpacing",
  zoom = "zoom"
}

export enum TextSettingsKeys {
  align = "align",
  fontFamily = "fontFamily",
  fontWeight = "fontWeight",
  hyphens = "hyphens",
  normalizeText = "normalizeText"
}

export const defaultTextSettingsMain = [TextSettingsKeys.fontFamily];

export const defaultTextSettingsSubpanel = [
  TextSettingsKeys.fontFamily,
  TextSettingsKeys.fontWeight,
  TextSettingsKeys.align,
  TextSettingsKeys.hyphens,
  TextSettingsKeys.normalizeText
]

export enum SpacingSettingsKeys {
  letterSpacing = "letterSpacing",
  lineHeight = "lineHeight",
  paraIndent = "paraIndent",
  paraSpacing = "paraSpacing",
  publisherStyles = "publisherStyles",
  wordSpacing = "wordSpacing"
}

export const defaultSpacingSettingsMain = [
  SpacingSettingsKeys.lineHeight
];

export const defaultSpacingSettingsSubpanel = [
  SpacingSettingsKeys.publisherStyles,
  SpacingSettingsKeys.lineHeight,
  SpacingSettingsKeys.paraSpacing,
  SpacingSettingsKeys.paraIndent,
  SpacingSettingsKeys.wordSpacing,
  SpacingSettingsKeys.letterSpacing
]

export enum SettingsRangeVariant {
  slider = "slider",
  numberField = "numberField"
}

export interface ISettingsTextPref {
  main?: TextSettingsKeys[];
  subPanel?: TextSettingsKeys[] | null;
  header?: SheetHeaderVariant;
}

export interface ISettingsSpacingPref {
  main?: SpacingSettingsKeys[];
  subPanel?: SpacingSettingsKeys[] | null;
  header?: SheetHeaderVariant;
  letterSpacing?: ISettingsRangePref;
  lineHeight?: {
    [key in Exclude<ReadingDisplayLineHeightOptions, ReadingDisplayLineHeightOptions.publisher>]: number
  };
  paraIndent?: ISettingsRangePref;
  paraSpacing?: ISettingsRangePref;
  wordSpacing?: ISettingsRangePref;
}

export interface ISettingsRangePref {
  variant?: SettingsRangeVariant;
  range?: [number, number];
  step?: number;
}

export const defaultParaSpacing: Required<ISettingsRangePref> = {
  variant: SettingsRangeVariant.slider,
  range: [0, 3],
  step: 0.5
}

export const defaultParaIndent: Required<ISettingsRangePref> = {
  variant: SettingsRangeVariant.slider,
  range: [0, 2],
  step: 0.25
}

export const defaultWordSpacing: Required<ISettingsRangePref> = {
  variant: SettingsRangeVariant.numberField,
  range: [0, 1],
  step: 0.1
}

export const defaultLetterSpacing: Required<ISettingsRangePref> = {
  variant: SettingsRangeVariant.numberField,
  range: [0, 0.5],
  step: 0.05
}

export const defaultLineHeights = {
  [ReadingDisplayLineHeightOptions.small]: 1.25,
  [ReadingDisplayLineHeightOptions.medium]: 1.5,
  [ReadingDisplayLineHeightOptions.large]: 1.75
}

export interface ISettingsMapObject {
  Comp: React.FC<IAdvancedDisplayProps> | React.ComponentType<any>;
  props?: any;
}

export interface IReadingDisplayGroupWrapperProps {
  heading: string;
  moreLabel: string;
  moreTooltip: string;
  onMorePressCallback: (e: PressEvent) => void;
  settingsMap: { [key in SpacingSettingsKeys]: ISettingsMapObject } | { [key in TextSettingsKeys]: ISettingsMapObject };
  prefs?: ISettingsTextPref | ISettingsSpacingPref;
  defaultPrefs: {
    main: TextSettingsKeys[] | SpacingSettingsKeys[];
    subPanel: TextSettingsKeys[] | SpacingSettingsKeys[];
  };
}

export interface IAdvancedIconProps {
  className?: string;
  ariaLabel: string;
  placement: TooltipProps["placement"];
  tooltipLabel: string;
  onPressCallback: (e: PressEvent) => void;
  isDisabled?: boolean;
}

export interface IAdvancedDisplayProps {
  standalone?: boolean;
}

export interface ISettingsSteppersProps {
  decrementIcon?: ComponentType<SVGProps<SVGElement>> | null;
  decrementLabel: string;
  incrementIcon?: ComponentType<SVGProps<SVGElement>> | null;
  incrementLabel: string;
}

interface ISettingsRangeProps {
  standalone?: boolean;
  className?: string;
  label: string;
  defaultValue: number;
  value: number;
  onChangeCallback: (value: number) => void;
  range: [number, number];
  step: number;
  format?: Intl.NumberFormatOptions;
}

export interface ISettingsNumberFieldProps extends ISettingsRangeProps {
  steppers: ISettingsSteppersProps;
  virtualKeyboardDisabled?: boolean;
  readOnly?: boolean;
}

export interface ISettingsSliderProps extends ISettingsRangeProps {}

export interface ISettingsSwitchProps {
  name?: string;
  className?: string;
  heading?: string;
  label: string;
  isSelected: boolean;
  onChangeCallback: (isSelected: boolean) => void;
}

export interface IRCSSSettings {
  align: ReadingDisplayAlignOptions | null;
  colCount: string;
  fontFamily: string | null;
  fontSize: number;
  fontWeight: number;
  hyphens: boolean | null;
  letterSpacing: number | null;
  lineLength: number | null;
  lineHeight: ReadingDisplayLineHeightOptions | null;
  layoutStrategy: RSLayoutStrategy;
  normalizeText: boolean;
  paginated: boolean;
  paraIndent: number | null;
  paraSpacing: number | null;
  publisherStyles: boolean;
  theme: ThemeKeys;
  wordSpacing: number | null;
}